### PR TITLE
Additions for experimenter usability

### PR DESCRIPTION
**Pull request by kimberscott**
Add date_created fields to user and demographic data. For users, this allows us to easily see when people are signing up, and check the impact of various recruitment efforts. For demographic data this lets us fetch the most recent data more easily, as well as seeing when changes were made.

I've made some minor formatting changes to the study detail view to make longer text fields a bit more readable.

Add some explanations to researchers about what they should put in the contact info field and the age cutoff fields.

Proposed solution for giving access to email addresses to users with can_read_usernames permissions only (separate basic/full serializers for user data, selected in view based on permissions).

### CODE REVIEW

**Comment by pattisdr** [comment]
hey @kimberscott, let me know when this is ready for review.

File: accounts/migrations/0034_auto_20170911_1832.py [hunk 1]
```diff
@@ -14,6 +14,7 @@ class Migration(migrations.Migration):
     operations = [
         migrations.AlterModelOptions(
             name='user',
-            options={'ordering': ['username'], 'permissions': (('can_create_users', 'Can Create User'), ('can_view_users', 'Can View User'), ('can_edit_users', 'Can Edit User'), ('can_remove_users', 'Can Remove User'), ('can_view_user_permissions', 'Can View User Permissions'), ('can_edit_user_permissions', 'Can Edit User Permissions'), ('can_read_all_user_data', 'Can Read All User Data'))},
+            options={'ordering': ['username'], 'permissions': (('can_create_users', 'Can Create User'), ('can_view_users', 'Can View User'), ('can_edit_users', 'Can Edit User'), ('can_remove_users', 'Can Remove User'), ('can_view_user_permissions', 'Can View User Permissions'), ('can_edit_user_permissions', 'Can Edit User Permissions'), ('can_read_all_user_data', 'Can Read All User Data'),
+('can_read_usernames', 'Can Read User Usernames'))},
```
Comment (pattisdr) [inline accounts/migrations/0034_auto_20170911_1832.py]: You shouldn't alter old migrations; you need to create new ones.  1) Undo the changes to this file. 2) You need to add 'can_read_usernames' to the[ User model permissions under the Meta class](https://github.com/CenterForOpenScience/lookit-api/blob/develop/accounts/models.py#L234); follow the patterns there.  3) Then run `python manage.py makemigrations`.  4. Commit new migration/model changes.

File: accounts/serializers.py [hunk 2]
```diff
@@ -106,10 +103,14 @@ class Meta:
...
+            'date_created',
         )
 
-    # class JSONAPIMeta:
-    #     included_resources = ['demographics', ]
+class FullUserSerializer(BasicUserSerializer):
+    
+    class Meta():
+        model = User
+        fields = BasicUserSerializer.Meta().fields + ('username',)
```
[the remark this answers is not in the document]
Comment (pattisdr) [reply accounts/serializers.py]: Also, consider adding tests for this. We don't have good test coverage in this django app, but we do have API tests.  A test to make sure that usernames only show if you have this permission, and do not show if you don't, would be useful because of the sensitivity of this field.

File: accounts/serializers.py [hunk 3]
```diff
@@ -51,19 +51,16 @@ class Meta:
             'state',
             'density',
             'extra',
+            'date_created',
```
Comment (pattisdr) [inline accounts/serializers.py]: Not sure what you need for your use case, but when I recently added `date_created` to the Response model, on the Response serializer, I changed the field name to 'created_on' to match what your randomizer was expecting.  
 `created_on = serializers.DateTimeField(read_only=True, source='date_created')`

**Comment by kimberscott** [comment]
Wow, thank you so much for the careful review & explanations - working on the changes now, and sorry to require so much oversight!

File: accounts/models.py [hunk 4]
```diff
@@ -127,6 +127,8 @@ class User(AbstractBaseUser, PermissionsMixin, GuardianUserMixin):
     email_new_studies = models.BooleanField(default=True)
     email_study_updates = models.BooleanField(default=True)
     email_response_questions = models.BooleanField(default=True)
+    
+    date_created = models.DateTimeField(auto_now_add=True)
```
[the remark this answers is not in the document]
Comment (kimberscott) [reply accounts/models.py]: That would be very helpful! But more important is being able to save this info going forward. It seems that the migration script I added here https://github.com/kimberscott/lookit-api/blob/develop/accounts/migrations/0036_add_user_demographic_date_created.py at least creates these fields so we have them for new users/demographic data.

File: accounts/serializers.py (hunk 3 above)
Comment (kimberscott) [reply accounts/serializers.py]: I saw that - thanks! I figured it might be simpler, all else equal, just to use the same date_created name but not switch it in the serializer, but am open to doing it either way. Nothing else on Lookit (in the ember-lookit-frameplayer/lookit-api) makes use of the date created info for users or demographic data, so we don't have anything we need to match, and it's all the same to me as I get our analysis code that uses the API working.

File: accounts/serializers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kimberscott) [reply accounts/serializers.py]: I have added some basic tests - not exhaustive but just to check we see the username field with this permission but not if user is just admin with can_read_all_user_data.

File: api/views.py [hunk 5]
```diff
@@ -115,11 +121,11 @@ def get_queryset(self):
         all_users = super().get_queryset()
         # Users with can_read_all_user_data permissions can view all active users via the API
         if self.request.user.has_perm('accounts.can_read_all_user_data'):
-            return all_users.filter(is_active=True)
+            return all_users.defer('middle_name').filter(is_active=True)
```
[the remark this answers is not in the document]
Comment (kimberscott) [reply api/views.py]: Thanks for catching that - that was from playing around with defer initially to see if that would work to omit usernames from responses. (Was deferring middle_name in case username wasn't dropped because it's used as an index.) Removed now!

File: accounts/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pattisdr) [reply accounts/models.py]: Hey @kimberscott, do we need to migrate over new users that have been added since we ran this participant migration a month+ ago?

File: studies/forms.py [hunk 6]
```diff
@@ -60,6 +60,8 @@ class Meta:
             'exit_url': "Specify the page where you want to send your participants after they've completed the study.",
             'short_description': 'Give your study a description here.',
             'long_description': 'Explain the purpose of your study here.',
+            'contact_info': 'This should give the name of the PI for your study, and an email address where the PI or study staff can be reached with questions. Format: PIs Name (contact: youremail@lab.edu)',
+            'criteria': 'Text shown to families - this is not used to actually verify eligibility.'
```
[the remark this answers is not in the document]
Comment (kimberscott) [reply studies/forms.py]: Changed this, also proposed simple fix to keep the shared fields consistent but it's a bit clunky, feel free to leave out!

File: accounts/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kimberscott) [reply accounts/models.py]: Yes! Ideally we'd do this as close as possible to switching lookit.mit.edu to the new site.

[the remark this answers is not in the document]
Comment (pattisdr) [reply accounts/models.py]: I think we only want to run this migration one more time, and we need to make sure we're coordinated before we run it.  If you need the date_created field added to the participant migration, we will probably wipe the prod db and rerun, at the same time pulling in all the new users.

Also, we want to hold off until you've verified that the migration has everything you need.  Working on some pieces on our end to get you logged into new lookit production via oauth.

**Comment by kimberscott** [comment]
I believe I've now addressed all the concerns here!

**Comment by pattisdr** [comment]
Thanks for the quick response, will review and merge shortly.

**Comment by kimberscott** [comment]
Ack! I just realized there are still issues with formatting of the studies_list page when page width is smaller - fixing now & will push. I'm sorry!

**Comment by pattisdr** [comment]
no problem, I'm switching to another task then, will come back to this later this afternoon.

**Comment by kimberscott** [comment]
Should be all set for real now. Apologies again!

File: accounts/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kimberscott) [reply accounts/models.py]: Ok, sounds good!

File: api/tests/test_users.py [hunk 7]
```diff
@@ -68,6 +71,30 @@ def testSuperusersCanViewAllUsers(self):
...
+        # User with specific 'can_view_usernames' permission can see usernames in user data
+        self.emailpermissionuser = G(User, is_active=True,  given_name="ResearcherEmail")
+        assign_perm('accounts.can_read_usernames', self.emailpermissionuser)
+        self.client.force_authenticate(user=self.emailpermissionuser)
+        api_response = self.client.get(self.url, content_type="application/vnd.api+json")
+        self.assertEqual(api_response.status_code, status.HTTP_200_OK)
+        userList = api_response.json()['data']
+        self.assertGreater(len(userList), 0) # View self
+        for u in userList:
+            self.assertIn('username', u['attributes'].keys())
```
Comment (pattisdr) [inline api/tests/test_users.py]: Hooray! Thank you for adding tests 👍

**Comment by pattisdr** [comment]
Nice work @kimberscott. Merging.
